src/acme/ui/CustomerListUI.java

```diff
@@ -0,0 +1,353 @@
+    jTable1.setBorder(javax.swing.BorderFactory.createEtchedBorder());
+    jTable1.setFont(new java.awt.Font("Lucida Grande", 0, 16)); 
+    jTable1.setModel(new javax.swing.table.DefaultTableModel(objects, cols) {
+        Class[] types = new Class [] {
+            java.lang.String.class, java.lang.String.class, java.lang.String.class, java.lang.Boolean.class, javax.swing.JButton.class
+        };
+        boolean[] canEdit = new boolean [] {
+            //need last one true so can click button
+            false, false, false, false, true
+        };
+
+        public Class getColumnClass(int columnIndex) {
+            return types [columnIndex];
+        }
+
+        public boolean isCellEditable(int rowIndex, int columnIndex) {
+            return canEdit [columnIndex];
+        }
+    });
+    
+    //needed for buttons in table
+    TableColumn editCol = jTable1.getColumnModel().getColumn(4);
+    JButton editButton = new JButton("edit");
+    editCol.setCellRenderer(new ButtonRenderer());;
+    editCol.setCellEditor(new ButtonEditor(new JTextField()));
+  
+    //more table properties
+    jTable1.setColumnSelectionAllowed(false);
+    jTable1.getTableHeader().setFont(new Font("Lucida Grande", 0, 18)); 
+    jTable1.setRowHeight(24);
+    jTable1.setRowMargin(2);
+    
+    jTable1.setSelectionBackground(new Color(93, 184, 202));
+    jTable1.setSelectionForeground(new java.awt.Color(0, 0, 0));
+    jTable1.setShowGrid(true);
+    jTable1.setPreferredScrollableViewportSize(jTable1.getPreferredSize());
+    jTable1.setFillsViewportHeight( true );
+    jScrollPane1 = new JScrollPane( jTable1 );
+    jScrollPane1.setBounds(25, 75, 600, 200);
+    add(jScrollPane1);
+    
+    //new customer button
+    JButton btnNewCustomer = new JButton("New Customer");
+    btnNewCustomer.setBounds(25, 280, 120, 40);
+    btnNewCustomer.setBackground(lightBlue);
+    btnNewCustomer.setOpaque(true);
+    btnNewCustomer.addActionListener(new ActionListener() {
+      public void actionPerformed(ActionEvent e) {
+        goToNewCustomerPage();
+      }
+    });
+    add(btnNewCustomer);
+  }
+  
+  public static void main(String [] args) {
+    AcmeUI acme = new AcmeUI();
+    acme.setPanel(new CustomerListUI());
+    acme.getCompany().setCurrentUser(new User());
+}
+  
+//navigation methods-----------------------------
+ public void goToEditPage(String customerName)
+ { 
+   customerName = customerName.substring(2);
+   Customer customerToPass = null;
+   for(int x = 0; x< this.customersArray.length; x++)
+   {
+     if(this.customersArray[x].getName().equals(customerName))
+     {
+       customerToPass = this.customersArray[x];
+     }
+   }
+   this.getAcmeUI().customerAddUpdate(customerToPass);
+ }
+ 
+ public void goToNewCustomerPage()
+ {
+   //go to new customer page
+   this.getAcmeUI().customerAddUpdate(null);
+ }
+//------------------------------------------------
+ 
+ public Object[][] getObjectsForTable(Customer[] listOfCustomers)
```

Comment:
This function breaks the program if the customer is missing any information. Can you wrap it in a try/catch for a NullPointerException so that the program doesn't halt if a customer is missing one of these elements?

Reply:
I added one inside the loop. Thanks for noticing that.